tree_unique_args/src/ast.rs

```diff
@@ -0,0 +1,204 @@
+use crate::{Expr, Expr::*, Id, Order};
+
+pub fn give_fresh_ids(expr: &mut Expr) {
+    let mut id = 1;
+    give_fresh_ids_helper(expr, 0, &mut id);
+}
+
+fn give_fresh_ids_helper(expr: &mut Expr, current_id: i64, fresh_id: &mut i64) {
+    match expr {
+        Loop(id, input, body) => {
+            let new_id = *fresh_id;
+            *fresh_id += 1;
+            *id = Id(new_id);
+            give_fresh_ids_helper(input, current_id, fresh_id);
+            give_fresh_ids_helper(body, new_id, fresh_id);
+        }
+        Let(id, arg, body) => {
+            let new_id = *fresh_id;
+            *fresh_id += 1;
+            *id = Id(new_id);
+            give_fresh_ids_helper(arg, current_id, fresh_id);
+            give_fresh_ids_helper(body, new_id, fresh_id);
+        }
+        Arg(id) => {
+            *id = Id(current_id);
+        }
+        Function(id, body) => {
+            let new_id = *fresh_id;
+            *fresh_id += 1;
+            *id = Id(new_id);
+            give_fresh_ids_helper(body, new_id, fresh_id);
+        }
+        Call(id, arg) => {
+            *id = Id(current_id);
+            give_fresh_ids_helper(arg, current_id, fresh_id);
+        }
+        _ => expr.for_each_child(move |child| give_fresh_ids_helper(child, current_id, fresh_id)),
```

Comment:
What does `move` do?

Reply:
`move` is a rust keyword that says "give this closure ownership of captured values"
I don't understand it completely but it makes this code type check

Reply:
`move` gives the lambda ownership of the variables that it closes over. Here I would assume that it's necessary because `fresh_id` is a mutable reference.